### PR TITLE
feat: align `publicPath` options with webpack

### PR DESCRIPTION
**Pull request by xc2**
## Summary

- [x] Add function type support to several `publicPath` options. They are:

1. `output.publicPath`
2. `entry.[name].publicPath`
3. `generator<asset>.publicPath`
4. `EntryPlugin` and `DynamicEntryPlugin`

- [x] Support `[hash]` and `[fullhash]` for `publicPath` of output and entry
- [x] Support full filename template args for `generator.publicPath`

## Checklist

- [x] Tests **updated** (or not required).
- [x] Documentation **updated** (or not required).

### CODE REVIEW

File: packages/rspack-dev-server/src/middleware.ts [hunk 1]
```diff
@@ -13,13 +13,26 @@ function etag(buf: any) {
...
+			compilation ? compilation.getPath(raw) : raw({ hash: "XXXX" }, undefined);
+	} else if (/\[(hash|fullhash)[:\]]/.test(raw)) {
+		return (compilation?: Compilation) =>
+			compilation ? compilation.getPath(raw) : raw.replace(/\/$/, "") + "/";
+	} else {
+		return () => raw.replace(/\/$/, "") + "/";
+	}
+}
+
 export function getRspackMemoryAssets(
```
Comment (xc2) [inline packages/rspack-dev-server/src/middleware.ts:30]: This function seems to be obsoleted.

I'm not gonna introduce breaking change in this PR so I didn't remove this function.

Instead, I just made an essential change to it to make tsc succeeded.

File: crates/rspack_plugin_asset/src/lib.rs [hunk 2]
```diff
@@ -393,7 +393,22 @@ impl ParserAndGenerator for AssetParserAndGenerator {
             .module_generator_options
             .and_then(|x| x.asset_public_path())
           {
-            let public_path = public_path.render(compilation, &filename);
+            let public_path = match public_path {
```
Comment (xc2) [inline crates/rspack_plugin_asset/src/lib.rs:396]: https://github.com/webpack/webpack/blob/7b80b2b18db66abca6feb7b02a9089aca4bc8186/lib/asset/AssetGenerator.js#L327-L338

**Comment by xc2** [comment]
bench!

**Review by ahabhgk — APPROVED** [review]
Thanks!

File: packages/rspack-dev-server/src/middleware.ts (hunk 1 above)
Comment (ahabhgk) [reply packages/rspack-dev-server/src/middleware.ts:30]: It's ok to remove this

**Pull request merged**